### PR TITLE
refactor(connlib): only call `onDisconnect` for unrecoverable errors

### PR DESCRIPTION
**Pull request by thomaseizinger**
Previously, we called `onDisconnect` in two kinds of situations:

- With an error when we wanted the clients to clear the token
- Without an error when the token was still valid (i.e. after a call to `disconnect` from the clients)

This is unnecessarily redundant. Firezone is designed to **not** have a state of "signed in but disconnected". Thus, every time connlib calls `disconnect`, we should clear the token and sign the user out.

At present, we only do this for errors with the control plane. Errors in the actual tunnel are only logged and we continue trying to use the tunnel. There are errors in the tunnel where we should also give up (i.e. TUN device gone, fatal IO error, etc). At present, those are not yet bubbled up but we will at some point. Once we have https://github.com/firezone/firezone/pull/3682, it will be much easier to create a type-safe contract that ensures we only disconnect on fatal errors.

### CODE REVIEW

**Review by jamilbk — COMMENTED** [review]
This PR will need some updates in Apple and Android clients. I can test / make those in this PR

**Comment by jamilbk** [comment]
- Updated Apple clients
- Updated Android clients
- Removed Option<> from Error since we should only call the callback when the error exists

## macOS demo

https://github.com/firezone/firezone/assets/167144/5ca085ad-f90d-4ea3-aa13-a50c1648bed8

## Android demo

https://github.com/firezone/firezone/assets/167144/e3f13307-806d-4ce7-a270-4a2de324572c

**Comment by jamilbk** [comment]
@ReactorScram Not sure if the Linux or Windows clients need any changes?

File: kotlin/android/app/src/main/java/dev/firezone/android/tunnel/callback/ConnlibCallback.kt [hunk 1]
```diff
@@ -25,8 +25,6 @@ interface ConnlibCallback {
     // The JNI doesn't support nullable types, so we need two method signatures
```
Comment (jamilbk) [inline kotlin/android/app/src/main/java/dev/firezone/android/tunnel/callback/ConnlibCallback.kt:25]:
```suggestion
```

**Comment by thomaseizinger** [comment]
> * Removed Option<> from Error since we should only call the callback when the error exists

Yeah? How do I now signal to the app that I gave retrying to connect to the portal and thus the tunnel is disconnected but the token may still be valid?

**Comment by jamilbk** [comment]
> > * Removed Option<> from Error since we should only call the callback when the error exists
> 
> 
> 
> Yeah? How do I now signal to the app that I gave retrying to connect to the portal and thus the tunnel is disconnected but the token may still be valid?

That takes 7 days I think currently.

There are three cases where this can happen:
1. device has no internet
2. device or firewall is blocking connection to portal
3. portal is down

(1) we can receive easily from the app layer and "sleep" connlib's control plane (i.e. just temporarily set the backoff super high)
(2) Is the tricky one, but expected to be rare. This is where onDisconnect with a different error type could be useful.
(3) is expected to be temporary

If we implement (1), we can lower MAX_PARTITION_TIME to a more reasonable value (say 15 minutes to match the data plane global timeout) and just sign the user out in this rare case. Today MAX_PARTITION_TIME is so high because (1) isn't implemented, otherwise it would annoyingly sign out clients with no internet (big problem for headless mode).

We have no "signed in but can't connect" state in the clients, and I'm not sure now's the time to add that complexity.

I think we can make this a good UX today with the following changes:

- Implement (1) above
- Use different MAX_PARTTITON_TIME for headless and user mode clients. (30 days for the former and 15 minutes for the latter)
- Call onDisconnect(error=connect_timeout) when MAX_PARTITION_TIME is hit

On rare cases users may have to sign in when they didn't have to, but considering they would have needed connectivity to the control plane to obtain the token in the first place I think this is acceptable.

In other words we try to avoid "signed in but can't connect" state. You're not really signed in if you can't connect.

In the future we make this nicer for GUI clients with the added state, and a button to "retry"

**Comment by jamilbk** [comment]
See #4026, #4027, and #4028

File: rust/gui-client/src-tauri/src/client/gui.rs [hunk 2]
```diff
@@ -451,18 +450,10 @@ enum CallbackError {
...
         tracing::debug!("on_disconnect {error:?}");
-        self.ctlr_tx.try_send(match error {
-            Some(connlib_client_shared::Error::ClosedByPortal) => {
-                // TODO: this can happen for other reasons
-                ControllerRequest::DisconnectedTokenExpired
-            }
-            _ => ControllerRequest::Disconnected,
-        })?;
+        self.ctlr_tx
+            .try_send(ControllerRequest::DisconnectedTokenExpired)?;
```
Comment (ReactorScram) [inline rust/gui-client/src-tauri/src/client/gui.rs]: So TokenExpired is the only reason we can get on_disconnect now? And that also covers cases where the client is purposely removed, so their token is not really expired, but invalid, or where they accidentally get the wrong token on disk? (say a backup and restore went bad, or a VM rolled back by accident)

And everything else like network issues, connlib will keep trying to stay connected and there's nothing the client needs to do?

Comment (jamilbk) [reply rust/gui-client/src-tauri/src/client/gui.rs]: Yep, that's correct

**Comment by ReactorScram** [comment]
The notification does work if I revoke my Windows client's token. Tested on 363c9b74c

Also I renamed `DisconnectedTokenExpired` to just `Disconnected` since there are no other disconnection types.

image

**Review by jamilbk — APPROVED** [review]

**Comment by thomaseizinger** [comment]
I can't approve my own PR but I've updated the PR description and to match my understanding of what and why we are doing this now. Please take a look and merge if you think it is correct.
